File: GuideObjC/GuideObjC/ViewControllers/PostController/CommentCell.m
//
//  CommentCell.m
//  GuideObjC
//
//  Created by LuzanovRoman on 19.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "CommentCell.h"
#import "EGFHumanName+Additions.h"

@interface CommentCell ()
@property (weak, nonatomic) IBOutlet UILabel *creatorNameLabel;
@property (weak, nonatomic) IBOutlet UILabel *descriptionLabel;
@property (weak, nonatomic) IBOutlet UIButton *deleteButton;
@property (weak, nonatomic) IBOutlet UIButton *editButton;
@end

@implementation CommentCell

+ (CGFloat)heightForComment:(EGFComment *)comment {
    CGFloat height = 47; // height of cell without text
    
    UIFont *font = [UIFont systemFontOfSize:15];
    CGFloat width = [UIScreen mainScreen].bounds.size.width - 32;
    
    if (comment.text) {
        CGSize size = CGSizeMake(width, FLT_MAX);
        CGRect frame = [comment.text boundingRectWithSize:size options:NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName: font} context:nil];
        height += frame.size.height + 8;
    }
    return height;
}

- (void)setComment:(EGFComment *)comment {
    _comment = comment;
    _deleteButton.hidden = ![[_delegate authorizedUserId] isEqual:comment.creator];
    _editButton.hidden = ![[_delegate authorizedUserId] isEqual:comment.creator];
    _creatorNameLabel.text = [comment.creatorObject.name fullName];
    _descriptionLabel.text = comment.text;
}

- (IBAction)deleteComment:(id)sender {
    [_delegate deleteComment:_comment];
}

- (IBAction)editComment:(id)sender {
    [_delegate editComment:_comment];
}
@end

File: GuideSwift/GuideSwift/ViewControllers/PostController/CommentCell.swift
//
//  CommentCell.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 19.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit

protocol CommentCellDelegate: NSObjectProtocol {
    var authorizedUserId: String? { get }
    func delete(comment: EGFComment)
    func edit(comment: EGFComment)
}

class CommentCell: UITableViewCell {
    @IBOutlet weak var creatorNameLabel: UILabel!
    @IBOutlet weak var descriptionLabel: UILabel!
    @IBOutlet weak var deleteButton: UIButton!
    @IBOutlet weak var editButton: UIButton!
    weak var delegate: CommentCellDelegate?

    static func height(forComment comment: EGFComment) -> CGFloat {
        var height: CGFloat = 51 // height of cell without text

        let font = UIFont.systemFont(ofSize: 15)
        let width = UIScreen.main.bounds.size.width - 32

        if let value = comment.text {
            let size = CGSize(width: width, height: CGFloat(FLT_MAX))
            let frame = value.boundingRect(with: size, options: .usesLineFragmentOrigin, attributes: [NSFontAttributeName: font], context: nil)
            height += frame.size.height + 8
        }
        return height
    }

    weak var comment: EGFComment? {
        didSet {
            if let userId = delegate?.authorizedUserId, let creatorId = comment?.creator, userId == creatorId {
                deleteButton.isHidden = false
                editButton.isHidden = false
            } else {
                deleteButton.isHidden = true
                editButton.isHidden = true
            }
            creatorNameLabel.text = comment?.creatorObject?.name?.fullName()
            descriptionLabel.text = comment?.text
        }
    }

    @IBAction func deleteComment(_ sender: AnyObject) {
        guard let theDelegate = delegate, let theComment = comment else { return }
        theDelegate.delete(comment: theComment)
    }

    @IBAction func editComment(_ sender: AnyObject) {
        guard let theDelegate = delegate, let theComment = comment else { return }
        theDelegate.edit(comment: theComment)
    }
}
